cmd/bucket-stats.go

```diff
@@ -19,10 +19,135 @@ package cmd
 
 import (
 	"sync/atomic"
+	"time"
 )
 
 //go:generate msgp -file $GOFILE
 
+// MeanMetric helps calculates an approximate
+// mean value with low memory overhead
+type MeanMetric struct {
+	Count int64
+	Mean  int64
+}
+
+// Add new value and update the mean
+func (m *MeanMetric) update(val int64) {
+	oldCount := atomic.LoadInt64(&m.Count)
+	oldMean := atomic.LoadInt64(&m.Mean)
+	// Approximate formula for generating mean value without a lot of memory
+	atomic.AddInt64(&m.Count, 1)
+	atomic.StoreInt64(&m.Mean, oldMean+(val-oldMean)/(oldCount+1))
+}
```

Comment:
Overall: I don't really like that this will become less and less responsive over time. So a server running for 2 weeks will take hours, maybe days to reflect a change in incoming data.

So I am not fond of this method of collecting metrics.

Secondly: You are treating 2 atomic values like a single atomic value. That makes it leaky since concurrent updates will miss data. You don't need oldCount, since `atomic.AddInt64` returns the new value and you +1 anyway. But I would like this to be done in another way.

Reply:
> Overall: I don't really like that this will become less and less responsive over time. So a server running for 2 weeks will take hours, maybe days to reflect a change in incoming data.

I am not I understand your point. What do you mean less responsive over time ? (by the way this is a realtime metric, not based on the scanner. So restarting the cluster will reset counters to zero)

Reply:
@vadmeste I am not sure you understand what it does.

`mean = oldMean+(val-oldMean)/(oldCount+1)`

This will take the difference to the current mean, divided by oldCount and adjust the mean by that.

As `oldCount` increases this difference will have less and less of an effect. If oldCount is 10000, you will need 10000 more samples to adjust it. If it is 1M it will need 1M samples and so on. This means that as time moves on your new data will have less and less of an effect.

You *can* reset it on reads. That will make reads modify state, but that may be acceptable if the period is close to constant.

Reply:
> As `oldCount` increases this difference will have less and less of an effect.

Oh yes.. you mean over time, the current mean value doesn't reflect the recent status.. Yeah I totally missed that. We probably need to use this, https://github.com/VividCortex/ewma,

Reply:
We already use this anyways @vadmeste we should use it.